In [ ]:
import wget
import os

urls = ["https://raw.githubusercontent.com/Sakaggi/dataengineeringtutorial/main/Extract/projects_data.csv", 
        "https://raw.githubusercontent.com/Sakaggi/dataengineeringtutorial/main/Extract/population_data.csv",
        "https://raw.githubusercontent.com/Sakaggi/dataengineeringtutorial/main/Extract/population_data.json",
        "https://raw.githubusercontent.com/Sakaggi/dataengineeringtutorial/main/Extract/population_data.xml"
        ]
def download_csv(urls):
    for url in urls:
        filename = url.split("/")[-1]
        output_path = "D:\@@\P"
        path = "{}".format(output_path+'\\'+filename)
        check_file = os.path.exists(path)
        if check_file:
            print(f"File downloaded: {output_path+filename}")
        else:
            try:
                filename = wget.download(url, out=output_path)
                print(f"File downloaded: {filename}")
            except Exception as e:
                print(f"An error occurred: {e}")
download_csv(urls)

In [2]:
import pandas as pd
# pd.set_option('display.max_rows', None)

df_projects = pd.read_csv("projects_data.csv", dtype= str) # sử dụng dtype=str vì có các cột có nhiều định đạng dữ liệu khác nhau mà dataframe ko thể xác định, do vậy đưa giá trị trong các cột trở thành str
df_projects

id                   regionname  \
0      P162228                        Other   
1      P163962                       Africa   
2      P167672                   South Asia   
3      P158768                   South Asia   
4      P161364                       Africa   
...        ...                          ...   
18243  P006578  Latin America and Caribbean   
18244  P037451      Europe and Central Asia   
18245  P037362      Europe and Central Asia   
18246  P037452      Europe and Central Asia   
18247  P037383      Europe and Central Asia   

                                             countryname prodline  \
0                                            World;World       RE   
1      Democratic Republic of the Congo;Democratic Re...       PE   
2      People's Republic of Bangladesh;People's Repub...       PE   
3      Islamic  Republic of Afghanistan;Islamic  Repu...       PE   
4      Federal Republic of Nigeria;Federal Republic o...       PE   
...                                                  ...      ...   
18243                Republic of Chile;Republic of Chile       PE   
18244  Grand Duchy of Luxembourg;Grand Duchy of Luxem...       PE   
18245              Kingdom of Denmark;Kingdom of Denmark       PE   
18246  Kingdom of the Netherlands;Kingdom of the Neth...       PE   
18247                    French Republic;French Republic       PE   

                       lendinginstr lendinginstrtype envassesmentcategorycode  \
0      Investment Project Financing               IN                        C   
1      Investment Project Financing               IN                        B   
2      Investment Project Financing               IN                      NaN   
3      Investment Project Financing               IN                        A   
4      Investment Project Financing               IN                        B   
...                             ...              ...                      ...   
18243      Specific Investment Loan               IN                      NaN   
18244    Structural Adjustment Loan               AD                      NaN   
18245    Structural Adjustment Loan               AD                      NaN   
18246    Structural Adjustment Loan               AD                      NaN   
18247    Structural Adjustment Loan               AD                      NaN   

      supplementprojectflg productlinetype projectstatusdisplay  ...  \
0                        N               L               Active  ...   
1                        N               L               Active  ...   
2                        Y               L               Active  ...   
3                        N               L               Active  ...   
4                        N               L               Active  ...   
...                    ...             ...                  ...  ...   
18243                    N               L               Closed  ...   
18244                    N               L               Closed  ...   
18245                    N               L               Closed  ...   
18246                    N               L               Closed  ...   
18247                    N               L               Closed  ...   

      mjtheme3name mjtheme4name mjtheme5name  \
0              NaN          NaN          NaN   
1              NaN          NaN          NaN   
2              NaN          NaN          NaN   
3              NaN          NaN          NaN   
4              NaN          NaN          NaN   
...            ...          ...          ...   
18243          NaN          NaN          NaN   
18244          NaN          NaN          NaN   
18245          NaN          NaN          NaN   
18246          NaN          NaN          NaN   
18247          NaN          NaN          NaN   

                                                location  \
0                                                    NaN   
1                                                    NaN   
2                      

In [3]:
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

- Có thể thấy rằng khi kiểm tra số giá trị NULL, số 18248 xuất hiện nhiều lần, nên ta có thể hiểu rằng có rất nhiều cột không có giá trị.
- Cột countryname không có giá trị null nào xuất hiện nhưng cột Country lại có 14045 giá trị null.
- Do vậy rõ ràng có vấn đề với bộ dữ liệu này.

In [4]:
df_projects.shape

(18248, 57)

- Sử dụng thuộc tính shape để xem số lượng (hàng, cột) của bộ dữ liệu.
- Kết quả cho thấy bảng có 18248 dòng, như vậy tồn tại rất nhiều dòng được lấp đầy bởi giá trị NULL, ta sẽ xử lý phần này ở bước Transform

In [6]:
df_population = pd.read_csv("population_data.csv", on_bad_lines= "skip", skiprows= 2, header= 1, index_col= None)
# on_bad_lines= "skip" : bỏ qua những dòng trống
# skiprows = 2: bỏ qua 2 dòng đầu
# header = 1: lấy dòng đầu tiên làm tiêu đề cột
# index_col = None: thiết lập giá trị index cho từng dòng, None trả về giá trị từ 0 tới len(data)
df_population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965  ...        2009  \
0    55438.0    56225.0    56695.0    57032.0    57360.0  ...    101453.0   
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0  ...  28004331.0   
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0  ...  22549547.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0  ...   2927519.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0  ...     84462.0   

         2010        2011        2012        2013        2014        2015  \
0    101669.0    102053.0    102577.0    103187.0    103795.0    104341.0   
1  28803167.0  29708599.0  30696958.0  31731688.0  32758020.0  33736494.0   
2  23369131.0  24218565.0  25096150.0  25998340.0  26920466.0  27859305.0   
3   2913021.0   2905195.0   2900401.0   2895092.0   2889104.0   2880703.0   
4     84449.0     83751.0     82431.0     80788.0     79223.0     78014.0   

         2016        2017  Unnamed: 62  
0    104822.0    105264.0          NaN  
1  34656032.0  35530081.0          NaN  
2  28813463.0  29784193.0          NaN  
3   2876101.0   2873457.0          NaN  
4     77281.0     76965.0          NaN  

[5 rows x 63 columns]

- Sử dụng dtypes để kiểm xem các kiểu dữ liệu của từng cột (opject là kiểu dữ liệu trộn lẫn nhiều kiểu khác nhau)

In [7]:
df_population.dtypes

Country Name       object
Country Code       object
Indicator Name     object
Indicator Code     object
1960              float64
                   ...   
2014              float64
2015              float64
2016              float64
2017              float64
Unnamed: 62       float64
Length: 63, dtype: object

- Tiếp theo kiểm tra xem dữ liệu trên từng hàng có bao nhiêu cột chứa giá trị NULL

In [8]:
# pd.set_option('display.max_rows', None)
df_population.isnull().sum(axis= 1)

0      1
1      1
2      1
3      1
4      1
      ..
259    1
260    1
261    1
262    1
263    1
Length: 264, dtype: int64

- Có thể thấy rằng toàn bộ các dòng đều có 1 giá trị null, ta tiếp tục kiểm tra xem số dòng trống trong mỗi cột

In [9]:
df_population.isnull().sum()

Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960                4
                 ... 
2014                2
2015                2
2016                2
2017                2
Unnamed: 62       264
Length: 63, dtype: int64

- Từ kết quả trên, thấy rằng 264 dòng đều có giá trị NULL mà cột Unnamed: 62 lại gồm 264 giá trị NULL, do đó cần loại bỏ giá trị này

In [11]:
df_population = df_population.drop(['Unnamed: 62'], axis= 1)

In [12]:
df_population.isnull().any(axis=1)

0      False
1      False
2      False
3      False
4      False
       ...  
259    False
260    False
261    False
262    False
263    False
Length: 264, dtype: bool

- Sau đó, kiểm tra các dòng tồn tại giá trị NULL

In [13]:
df_population[df_population.isnull().any(axis=1)]

Country Name Country Code     Indicator Name Indicator Code  \
67                     Eritrea          ERI  Population, total    SP.POP.TOTL   
108             Not classified          INX  Population, total    SP.POP.TOTL   
125                     Kuwait          KWT  Population, total    SP.POP.TOTL   
194         West Bank and Gaza          PSE  Population, total    SP.POP.TOTL   
212                     Serbia          SRB  Population, total    SP.POP.TOTL   
223  Sint Maarten (Dutch part)          SXM  Population, total    SP.POP.TOTL   

          1960       1961       1962       1963       1964       1965  ...  \
67   1397491.0  1432640.0  1469645.0  1508273.0  1548187.0  1589179.0  ...   
108        NaN        NaN        NaN        NaN        NaN        NaN  ...   
125   269618.0   301336.0   338296.0   379891.0   425235.0   473554.0  ...   
194        NaN        NaN        NaN        NaN        NaN        NaN  ...   
212        NaN        NaN        NaN        NaN        NaN        NaN  ...   
223        NaN        NaN        NaN        NaN        NaN        NaN  ...   

          2008       2009       2010       2011       2012       2013  \
67   4232636.0  4310334.0  4390840.0  4474690.0        NaN        NaN   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125  2652340.0  2818939.0  2998083.0  3191051.0  3395556.0  3598385.0   
194  3596688.0  3702218.0  3811102.0  3927051.0  4046901.0  4169506.0   
212  7350222.0  7320807.0  7291436.0  7234099.0  7199077.0  7164132.0   
223    33964.0    34238.0    34056.0    33435.0    34640.0    36607.0   

          2014       2015       2016       2017  
67         NaN        NaN        NaN        NaN  
108        NaN        NaN        NaN        NaN  
125  3782450.0  3935794.0  4052584.0  4136528.0  
194  4294682.0  4422143.0  4551566.0  4684777.0  
212  7130576.0  7095383.0  7058322.0  7022268.0  
223    37685.0    38824.0    39969.0    41109.0  

[6 rows x 62 columns]

- Từ kết quả cho thấy, bộ dữ liệu về population chỉ có 6 giá trị bị thiếu. Ta sẽ xử lý trong phần Transform.

In [ ]:
df_json = pd.read_json("population_data.json", orient="records")
df_json